### ISSUE TITLE
First Broker Flow Allows Setting Arbitrary User Attributes

### ISSUE THREAD

**Issue by abstractj** [keycloak/keycloak#26111]
### Description

When a user authenticates using an Identity Provider for the first time, the "First Broker Login" flow is triggered (default behavior, can be configured). This flow requires the user to configure his email, first and last name. However, this flow also allows users to define user attributes by prefixing the key with "user.attributes.".

Setting such attributes for a user is expected to be an action only administrators can perform. These do not appear to be used to dictate access controls.

### Version
\>= 23.0.4

### Impact

Improper authorization that allows a non admin user to define user attributes on his user.

### Recommendations

Review the application logic and/or use-case for allowing users to configure attributes on their accounts upon import from an Identity Provider.

### References: 
- https://issues.redhat.com/browse/RHSSO-2687

**Comment by mposolda** [comment]
@SecKatie @abstractj Just a note that:

Since Keycloak 24, the user profile would be enabled by default. And with user profile enabled, custom attributes are not allowed to be used by default. The only exception is, when `Unmanaged attributes` switch on the Realm Settings page would be set to `Enabled` - in that case the arbitrary attributes are still allowed because of the backwards compatibility. But note that for all those cases, we have read-only attribute settings, which allows to specify attributes, which should not be editable by regular users or by administrators https://www.keycloak.org/docs/latest/server_admin/index.html#read_only_user_attributes . Administrator is supposed to add here all the "metadata attributes" he uses or all the attributes, which are possibly used for ABAC etc.

Also note that this problem is not specific to identity brokering, but exists in various other places as well (like account REST API, admin REST API, regular update-profile page etc). We had CVE for this in the past, which we addressed exactly by introducing read-only attributes.

So in summary: The recommended way is to have user-profile enabled with `Unmanaged attributes` set to `Disabled` (or eventually to `Only admin can edit/view` which allows only administrator to setup random attributes, but regular users won't be allowed to setup random attributes). If `Unmanaged attributes` need to be kept to `Enabled`, then the option is to rely on read-only attributes, where admin need to specify the attributes. We should document this in the Keycloak 24 documentation clearly.

Would it be sufficient to handle this issue this way and just consider as docs issue?

**Comment by sschu** [comment]
Wouldn't this be handled by the user profile when it is enabled by default? Or is the first broker login flow modifying the user directly bypassing the user profile?

**Comment by pedroigor** [comment]
```
If Unmanaged attributes need to be kept to Enabled, then the option is to rely on read-only attributes, where admin need to specify the attributes. We should document this in the Keycloak 24 documentation clearly.
```

IMO, the recommendation should be to set `ADMIN_EDIT|VIEW` and define all user-facing attributes in the user profile configuration?

But yeah, I think this is something to add to docs. Created link to this issue https://github.com/keycloak/keycloak/issues/26154.

**Comment by sschu** [comment]
@mposolda I just didn't see your comment, probably the page was open from before you wrote it.

**Comment by mposolda** [comment]
> @mposolda I just didn't see your comment, probably the page was open from before you wrote it.

@sschu ah, ok. Looks we were commenting almost around same time to this issue :-)

**Comment by abstractj** [comment]
> Would it be sufficient to handle this issue this way and just consider as docs issue?

That makes perfect sense @mposolda and the link created by Pedro also helps us to keep track of those things. Thank you!

**Comment by sguilhen** [comment]
@pedroigor Can this be closed? It is mentioned in https://github.com/keycloak/keycloak/issues/26154 as a task and that issue was closed, so I imagine the task was completed, no?
